### PR TITLE
Added calculation of scores above 999,999 for any tetris version

### PR DESCRIPTION
**Pull request by Torbaz**
Calculated and added score when maxout is reached. (When 999,999 is on scoreboard and another line is completed)

Added node_modules and .env to the .gitignore

### CODE REVIEW

**Comment by alex-ong** [comment]
Looks like this check will skip the 7~ frame image glitch; I suspect it's better to move this code to inside the check. I'll write some comments inside the code

File: public/views/main.js [hunk 1]
```diff
@@ -284,14 +284,14 @@ function onFrame(event, debug) {
...
+	// check for score change or score stayed at 999999 but the line count changed.
+	if (--pending_line === 0 || (diff.cleared_lines > 0 && transformed.score == 999999)) {
 		if (
 			transformed.score
 			&& (transformed.lines != null)
-			&& (transformed.level != null)
-			&& diff.score >= 0
-			&& diff.cleared_lines >= 0
+			&& (transformed.level != null) 
+			&& ((diff.score >= 0 && diff.cleared_lines >= 0)
```
Comment (alex-ong) [inline public/views/main.js:293]: Consider a function for this? Such as
ScoreStuckOnRollover or such

File: public/views/main.js [hunk 2]
```diff
@@ -284,14 +284,14 @@ function onFrame(event, debug) {
 		}
 	}
 
-	// check for score change
-	if (--pending_line === 0) {
+	// check for score change or score stayed at 999999 but the line count changed.
+	if (--pending_line === 0 || (diff.cleared_lines > 0 && transformed.score == 999999)) {
```
Comment (alex-ong) [inline public/views/main.js:288]: Is this check required here? Since the same check appears is below

If it's here then it could go the if statement nside when pendingline is !== 1 which could be disastrous as you could read the wrong amount of lines cleared.

**Review by alex-ong — COMMENTED** [review]
I pressed review instead of comment >.<

File: public/views/main.js [hunk 3]
```diff
@@ -284,14 +284,14 @@ function onFrame(event, debug) {
...
 		if (
 			transformed.score
 			&& (transformed.lines != null)
-			&& (transformed.level != null)
```
Comment (alex-ong) [inline public/views/main.js:292]: I know this is yobis code here but since this is 5 sets of && || could this be compressed into isValidDiff or such

**Review by alex-ong — COMMENTED** [review]
This is yobis code but feel free to " shorten"

File: public/views/main.js (hunk 2 above)
Comment (Torbaz) [reply public/views/main.js:288]: Thank you for this I'll take a look when I'm home

**Comment by Torbaz** [comment]
Fixed the bypassing of the pending_line check. I'm just working on getting my head around how this all works so it's taking a while to make these changes.

**Comment by alex-ong** [comment]
Looks good! My only concern is if they clear a line and single to land them at 999999 exactly does it work as intended or not (my guess is that it works). But apart from that one case, you appear to be using the 7 frame magic noise buffer correctly, so if the behavior is correct then yobi can merge

**Comment by Torbaz** [comment]
As far as I know, this does correctly handle cases where the score is exactly at 999,999. As it checks the event.score (The score read from the ocr) to be 999,999 and adds the correct score to the score from the last frame, so this should still work as all of these don't change when the score is 999,999. Thank you for the advice with the 7 frame check.

File: public/views/main.js [hunk 4]
```diff
@@ -284,14 +284,14 @@ function onFrame(event, debug) {
...
 	if (--pending_line === 0) {
 		if (
 			transformed.score
 			&& (transformed.lines != null)
-			&& (transformed.level != null)
-			&& diff.score >= 0
-			&& diff.cleared_lines >= 0
+			&& (transformed.level != null) 
+			&& ((diff.score >= 0 && diff.cleared_lines >= 0)
+			|| (transformed.score == 999999 && diff.cleared_lines > 0))
```
Comment (timotheeg) [inline public/views/main.js:294]: Was this addition really needed? 🤔 . If `score` is stuck at 999,999, the `diff.score` will be zero, and clearing lines will be greater than 0, so it seems the previous conditional was enough, no?

**Review by timotheeg — COMMENTED** [review]
Thanks for giving comments and feedback in the PR @alex-ong ! 🙏 

The code looks good, I just added one comment. 

On the making functions or aggregation booleans, `main.js` needs quite a lot of TLC cleanups all over the place, so I think that can come later.

@Torbaz Thanks for the PR. Just FYI, the layouts are divided into 2 family of processors. `main.js` is used by the `classic` and `das_trainer` layouts, while all other layouts use the `Player` class.  I intended to unify everything, likely using the Player class as the base for everything, but I haven't done it (maybe some day?). So the counting past 999,999 will need to be ported to be player class too at some point... (or maybe make the player class use `Game.js`). That unification is on the the cleanup TODO list.

One reason I haven't done it is that the layouts based on the player class show much less stats than the `classic` and `das_trainer` layouts. In particular the competition layouts show limited stats, so it felt a little wasteful to track so much thing about each game for these cases. But maybe it's not that much anyway 🤷‍♂️

File: public/views/main.js (hunk 4 above)
Comment (Torbaz) [reply public/views/main.js:294]: The reasoning for this change is the way I have implemented the score changing system is that it modifies the score saved in memory, but the transformed.score comes in every frame still at 999,999 even if the score is over that. This leads to a below 0 diff.score

I didn't know that about how the layouts are handled.
I will manually add this code to the player class for now, but once I have my head around this system some more I am happy to help with the porting of code to a unified solution.

**Comment by Torbaz** [comment]
I've implemented similar detection and calculation in the player class. Everything should be working as intended and I have tested it on https://www.youtube.com/watch?v=eIa1KHkzaXo

File: public/views/main.js (hunk 4 above)
Comment (timotheeg) [reply public/views/main.js:294]: > https://github.com/timotheeg/nestrischamps

Oh right, calling `game.onLine(evt)` mutates the event object, so by the time it's stored in the last state of main, it's the incremented value of the score.

I'm actually thinking the event doesn't need to be mutated 🤔 . The score is displayed from the game data:
```
dom.score.current.textContent = game.data.score.current.toString().padStart(6, '0').padStart(7, ' ');
```
so as long as the data internal to the game instance is tracked correctly, the event and state can be treated as raw data, and the main file doesn't need to worry about the cap at 999,999. 

I think merge the PR first, but I might modify that later.


> I didn't know that about how the layouts are handled.

With dirty code and zero documentation, there was no way you could have known 😅 .


> I will manually add this code to the player class for now

Thank you! 🙏  I would have been happy with just main supporting this first, while implementation for player could be delayed. Thanks for jumping into it right away!
